balancer_conn_wrappers.go

```diff
@@ -92,16 +115,19 @@ func (ccb *ccBalancerWrapper) updateClientConnState(ccs *balancer.ClientConnStat
 		}
 		errCh <- ccb.balancer.UpdateClientConnState(*ccs)
 	})
-
-	// If the balancer wrapper is closed when waiting for this state update to
-	// be handled, the callback serializer will be closed as well, and we can
-	// rely on its Done channel to ensure that we don't block here forever.
-	select {
-	case err := <-errCh:
-		return err
-	case <-ccb.serializer.Done:
-		return nil
+	if !ok {
+		// If we are unable to schedule a function with the serializer, it
+		// indicates that it has been closed. A serializer is only closed when
+		// the wrapper is closed or is in idle.
+		ccb.mu.Unlock()
+		return fmt.Errorf("grpc: cannot send state update to a closed or idle balancer")
```

Comment:
This should only happen if a balancer that was already `Closed` calls, right?

If so, I think we should `logger.Error` this.

Reply:
This is making me think...

What happens if an LB policy (or resolver; I haven't looked at `resolver_conn_wrapper.go` yet) is closed due to entering idle, then the channel exits idle, and then the old LB policy (or resolver) makes calls on the `ClientConn`.  It seems we need to either create new wrappers instead of reusing the existing ones, OR we need to ensure that `ccb.balancer == <the balancer calling us but I don't know how we could even check this>`, OR we need to make a `ccBalancerWrapperWrapper` that is invalidated when the LB policy is closed (this is hopefully not a serious suggestion).

Especially since we delay closure by calling it in a goroutine, it seems like this is a very real possibility.

Reply:
Based on our offline discussion, I switched the resolver wrapper to be closed on entering idle and recreated upon exiting idle.

But for the balancer and picker wrappers, I left it as is, i.e with methods to enter and exit idle. For the balancer wrapper, once the underlying balancer is closed, the graceful switch balancer (which is always our top-level balancer) will take care of dropping calls from the old (and closed) LB policy. 

Also, for the balancer and picker wrappers, it is not be easy to close and recreate them because that would mean that all accesses to them would need to be guarded with `cc.mu`. This is not possible because the picker wrapper needs to be accessed at pick time, while the balancer wrapper needs to be accessed whenever there is a subConn state change and when `cc.Connect` is called.

Reply:
I was thinking the picker wrapper would be safe to access, since you would do so only after going through the idleness manager, which should ensure that the picker wrapper exists before returning.  And a mutex around balancer accesses doesn't seem too troublesome.  It would be nice to handle these two subcomponents the same way, but I'm fine with this approach if it's working and gets us idleness working.

Reply:
I agree that it would be nicer to follow the same approach with the other two wrappers. But given that I'm constrained for time at this point, I have left a TODO for the same.